rundeckapp/grails-app/services/rundeck/services/asyncimport/AsyncImportService.groovy

```diff
@@ -0,0 +1,1073 @@
+
+        asyncImportStatusFileUpdater(new AsyncImportStatusDTO(projectName, milestoneNumber).with {
+            it.lastUpdate = "Creating the working directory in /tmp."
+            return it
+        })
+
+        String scopedWorkingDir = "${BASE_WORKING_DIR}${projectName}"
+        File baseWorkingDir = new File(scopedWorkingDir)
+        File modelProjectHost = new File(baseWorkingDir.toString() + File.separator + MODEL_PROJECT_NAME_SUFFIX)
+        createDirs(List.of(
+                baseWorkingDir,
+                modelProjectHost
+        ))
+
+        logger.debug("Creating the project model inside working directory in /tmp.")
+
+        asyncImportStatusFileUpdater(new AsyncImportStatusDTO(projectName, milestoneNumber).with {
+            it.lastUpdate = "Creating the project model inside working directory in /tmp."
+            return it
+        })
+
+        def framework = frameworkService.rundeckFramework
+        Predicate<? super Path> internalProjectMatcher = path -> path.fileName.toString().startsWith(MODEL_PROJECT_INTERNAL_PREFIX)
+
+        try {
+            if( modelProjectHost.list().size() == 0 ){
+
+                logger.debug("Checking if there are executions in project.")
+
+                // check if the executions dir exists, if not, false a flag to prevent M2 trigger
+                def internalProjectPath = getPathWithLogic(Paths.get(destDir.toString()), internalProjectMatcher)
+                if( !checkExecutionsExistenceInPath(internalProjectPath) ){
+                    executionsDirFound = false
+                }
+                copyDirExcept(destDir, modelProjectHost.toString(), EXECUTION_DIR_NAME)
+            }
+
+            def internalProjectInModelPath = getPathWithLogic(Paths.get(modelProjectHost.toString()), internalProjectMatcher)
+
+            if( internalProjectInModelPath != null ){
+                Files.move(internalProjectInModelPath, Paths.get(modelProjectHost.toString()).resolve("${MODEL_PROJECT_INTERNAL_PREFIX}${projectName}"))
+            }
+
+            String zippedFilename = "${baseWorkingDir.toString()}${File.separator}${projectName}${MODEL_PROJECT_NAME_EXT}"
+
+            if( !Files.exists(Paths.get(zippedFilename)) ){
+                try(FileOutputStream fos = new FileOutputStream(zippedFilename)){
+
+                    ZipOutputStream zos = new ZipOutputStream(fos)
+
+                    logger.debug("Ziping uploaded project.")
+
+                    zipDir(modelProjectHost.toString(), "", zos)
+
+                }catch(IOException ignored){
+                    ignored.printStackTrace()
+                    throw ignored
+                }
+            }
+
+            logger.debug("Uploading project w/o executions.")
+
+            asyncImportStatusFileUpdater(new AsyncImportStatusDTO(projectName, milestoneNumber).with {
+                it.lastUpdate = "Uploading project w/o executions."
+                return it
+            })
+
+            try {
```

Comment:
wouldn't be better to use "withCloseable" instead? to ensure no stream is left open